service_container/calls.rst

```diff
@@ -71,3 +71,74 @@ To configure the container to call the ``setLogger`` method, use the ``calls`` k
 
         $container->register(MessageGenerator::class)
             ->addMethodCall('setLogger', [new Reference('logger')]);
+
+When it comes to optional dependencies, the setter approach can be useful but
+what if you need to change the injected dependency without mutating the initial service?
+
+To do so, you probably could use `immutable-setter` injection::
+
+    namespace App\Service;
+
+    use Psr\Log\LoggerInterface;
+
+    class MessageGenerator
+    {
+        private $logger;
+
+        /**
+         * @return static
+         */
+        public function withLogger(LoggerInterface $logger)
+        {
+            $new = clone $this;
+            $new->logger = $logger;
+
+            return $new;
+        }
+
+        // ...
+    }
+
+In order to use this approach fully, just adapt the container configuration:
```

Comment:
We banned words like "just" recently. Here is an alternative proposal:

Because the method returns a separate cloned instance, configuring such a service means using the return value of the wither method (`$service = $service->withLogger($logger);`). The configuration to tell the container it should do so would be like:

Reply:
Improved, to check :)